### PR TITLE
Add forecastio dependency on requirements for weather app

### PR DESCRIPTION
**Pull request by admiral96**

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
 from flask import Flask
+from .api_parser import WeatherForecast
 
 
 mulungwishi_app = Flask(__name__)
+forecast = WeatherForecast()
```
Comment (mknecht) [inline app/__init__.py]: why?

File: test_url_runner.py [hunk 2]
```diff
@@ -12,6 +12,9 @@ def setUp(self):
         self.client = mulungwishi_app.test_client()
         self.client.testing = True
 
+
+    ''' test for mulungwishi '''
```
Comment (mknecht) [inline test_url_runner.py]: instead, use file names to make clear what you test.

File: app/views.py [hunk 3]
```diff
@@ -33,3 +49,7 @@ def page_forbidden(error):
 @url.errorhandler(500)
 def page_server_error(error):
     return render_template('500.html', title='Server Error'), 500
+
+
+def convert_to_celsius(temperature):
+    return (temperature - 32) * 5 / 9
```
Comment (mknecht) [inline app/views.py]: … really? :p

File: app/views.py [hunk 4]
```diff
@@ -20,6 +20,22 @@ def show_user_input():
...
+    if place:
+        weather_forecast = forecast.generate_forecast(address=place)
+        if isinstance(weather_forecast, dict):
+            temperature = convert_to_celsius(weather_forecast['currently']['temperature'])
+            summary = weather_forecast['currently']['summary']
+            humidity = weather_forecast['currently']['humidity']
+            address = weather_forecast['place_info']['formatted_address']
+            address_type = weather_forecast['place_info']['place_type']
+            return 'WEATHER FORECAST FOR {}: \nPlace Type: {}\nTemperature: {}°C\nHumidity: {}\nWeather Summary: {}'.format(address.upper(), address_type, round(temperature, 2), humidity, summary)
+        return '{}. Try a more generic place name'.format(weather_forecast), 400
```
Comment (mknecht) [inline app/views.py]: can you add an example?

File: app/views.py [hunk 5]
```diff
@@ -20,6 +20,22 @@ def show_user_input():
     return valid_query_message if content and sms_from and sms_to else (invalid_query_message, invalid_query_status_code)
 
 
+@url.route('/ask_weather')
+def generate_forecast():
+    place = request.args.get('address') or None
+    if place:
```
Comment (mknecht) [inline app/views.py]: please convert these nested ifs to an exit-early style.

File: app/api_parser.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        if place_query_status == 'OK':
+            place_type = geocodes['results'][0]['address_components'][0]['types'][0]
+            formatted_address = geocodes['results'][0]['formatted_address']
+            latitude = geocodes['results'][0]['geometry']['location']['lat']
+            longitude = geocodes['results'][0]['geometry']['location']['lng']
+            forecasts = ForeCast(latitude, longitude).get_request()
+            place_info = {'place_type': place_type, 'formatted_address': formatted_address}
+            forecasts.update({'place_info': place_info})
+            return forecasts
+        return str(place_query_status)
```
Comment (mknecht) [inline app/api_parser.py]: what is an example for the status?

File: app/api_parser.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        self.url = 'https://api.forecast.io/forecast/{}/{},{}'.format(FORECAST_API_KEY, latitude, longitude)
+
+    def get_request(self):
+        return json.loads((requests.get(self.url)).text)
+
+
+class GeoCode:
+
+    def __init__(self, address):
+        self.url = 'https://maps.googleapis.com/maps/api/geocode/json?address={}&key={}'.format(address, GOOGLE_API_KEY)
```
Comment (mknecht) [inline app/api_parser.py]: are you sure you don't need to url-encode the address? what if i text in “Akazienstraße 1” ? (note the sharp s)

File: app/api_parser.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import requests
+from config import FORECAST_API_KEY, GOOGLE_API_KEY
+
+
+class ForeCast:
+
+    def __init__(self, latitude, longitude):
+        self.url = 'https://api.forecast.io/forecast/{}/{},{}'.format(FORECAST_API_KEY, latitude, longitude)
+
+    def get_request(self):
```
Comment (mknecht) [inline app/api_parser.py]: use a property! :)

File: app/api_parser.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        return json.loads((requests.get(self.url)).text)
+
+
+class GeoCode:
+
+    def __init__(self, address):
+        self.url = 'https://maps.googleapis.com/maps/api/geocode/json?address={}&key={}'.format(address, GOOGLE_API_KEY)
+
+    def get_request(self):
```
Comment (mknecht) [inline app/api_parser.py]: property! pythonic!

File: config.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+import os
+
+
+GOOGLE_API_KEY = os.getenv('GOOGLE_API_KEY')
+FORECAST_API_KEY = os.getenv('FORECAST_API_KEY')
```
Comment (mknecht) [inline config.py]: nice!

**Comment by mknecht** [comment]
@admiral96 in a separate PR, make sure that the app returns a 500, when it's being started without the necessary API keys. Also, can you make sure the correct thing happens, when it can't reach the APIs?

File: app/views.py [hunk 11]
```diff
@@ -1,4 +1,4 @@
-from app import mulungwishi_app as url
+from app import mulungwishi_app as url, forecast
```
Comment (mknecht) [inline app/views.py]: alphabetically, please (use isort :) )

File: app/api_parser.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        return json.loads((requests.get(self.url)).text)
```
Comment (mknecht) [inline app/api_parser.py]: remove parens

File: app/api_parser.py (hunk 7 above)
Comment (mknecht) [inline app/api_parser.py]: make this a class attribute — no need to store per instance, right?

File: app/api_parser.py (hunk 6 above)
Comment (admiral96) [inline app/api_parser.py]: @mknecht OK and ZERO_RESULTS

File: app/api_parser.py (hunk 7 above)
Comment (admiral96) [inline app/api_parser.py]: I made a query for “Akazienstraße 1”:

WEATHER FORECAST FOR AKAZIENSTRASSE 1, 06366 KÖTHEN (ANHALT), GERMANY: Place Type: street_number Temperature: 4.56°C Humidity: 0.93 Weather Summary: Light Rain

How's it?

**Comment by admiral96** [comment]
Fail build on travis CI for main repo is natural for travis doesn't allow secure variables(environment variables) if the pull request hasn't merged yet. Build passed on forked repo.
Reference:
https://docs.travis-ci.com/user/encryption-keys/

**Comment by mknecht** [comment]
:+1:

**Pull request merged**